File: 25_2/s_15/exercise5/problem2.py
def csc_to_adj(col_pointers, row_indices, n, directed=True):
    adj = [[0 for _ in range(n)] for _ in range(n)]
    for j in range(n):
        start = col_pointers[j]
        end = col_pointers[j + 1]
        for k in range(start, end):
            i = row_indices[k]
            adj[i][j] = 1
            if not directed:
                adj[j][i] = 1
    return adj

def print_matrix(mat, labels=None):
    n = len(mat)
    if labels is None:
        labels = [str(i) for i in range(n)]
    print("   " + " ".join(labels))
    for i in range(n):
        row_str = " ".join(str(x) for x in mat[i])
        print(labels[i], row_str)

vertices = ["A", "B", "C", "D", "E"]
n = 5

col_pointers_1 = [0, 2, 5, 8, 11, 12]
row_indices_1 = [1, 2, 0, 2, 3, 0, 1, 3, 1, 2, 4, 3]

adj1 = csc_to_adj(col_pointers_1, row_indices_1, n, directed=False)
print("Graph 1 (undirected) adjacency matrix:")
print_matrix(adj1, vertices)
print()

col_pointers_2 = [0, 0, 2, 4, 5, 7]
row_indices_2 = [0, 3, 0, 1, 2, 1, 3]

adj2 = csc_to_adj(col_pointers_2, row_indices_2, n, directed=True)
print("Graph 2 (directed) adjacency matrix:")
print_matrix(adj2, vertices)
print()
print("Unique directed cycle in Graph 2: B -> C -> D -> B")

File: 25_1/s_11/ex5/problem2.c
/// Computer Science
/// Student Number is 11

#include <stdio.h>
#include <string.h>

#define N 5 // number of vertices A,B,C,D,E

void buildAdjMatrix(int n, int col_ptr[], int row_idx[], int directed, int adj[][N])
{
    // initialize the matrix with zeros
    memset(adj, 0, sizeof(int) * N * N);

    // reconstruct from CSC
    for (int col = 0; col < n; col++)
    {
        int start = col_ptr[col];
        int end = col_ptr[col + 1];

        for (int k = start; k < end; k++)
        {
            int row = row_idx[k];
            adj[row][col] = 1; // edge row -> col

            if (!directed)
            {
                adj[col][row] = 1; // make symmetric for undirected graph
            }
        }
    }
}

void printMatrix(int n, int adj[][N])
{
    printf("    A B C D E\n");
    for (int i = 0; i < n; i++)
    {
        printf("%c [ ", 'A' + i);
        for (int j = 0; j < n; j++)
        {
            printf("%d ", adj[i][j]);
        }
        printf("]\n");
    }
    printf("\n");
}

int main()
{

    // -----------------------------
    // Graph 1 (undirected)
    // -----------------------------
    int col1[] = {0, 2, 5, 8, 11, 12};
    int row1[] = {1, 2, 0, 2, 3, 0, 1, 3, 1, 2, 4, 3};

    int adj1[N][N];
    buildAdjMatrix(N, col1, row1, 0, adj1);

    printf("Adjacency Matrix: Graph 1 (Undirected)\n");
    printMatrix(N, adj1);

    // -----------------------------
    // Graph 2 (directed)
    // -----------------------------
    int col2[] = {0, 0, 2, 4, 5, 7};
    int row2[] = {0, 3, 0, 1, 2, 1, 3};

    int adj2[N][N];
    buildAdjMatrix(N, col2, row2, 1, adj2);

    printf("Adjacency Matrix: Graph 2 (Directed)\n");
    printMatrix(N, adj2);

    return 0;
}
